### PR TITLE
Enhancement: Houdini use instance transient data for `instance_node`

### PR DESCRIPTION
**Pull request by BigRoy**
## Changelog Description

Use transient data to store the actual houdini node object in the instance as opposed to the path so `hou.node()` isn't needed to retrieve the node per plugin.

Also removed:
- Instance toggled callback which does nothing with new publisher
- Remove legacy collect instances plugin which is not used for new publisher

## Additional notes

👇 **Please consider this question**

I wasn't sure whether I should open this as a Draft PR or not. I'm not sure whether everyone agrees this is a design improvement or whether it's worse. Thoughts?

☝️ 

## Testing notes:

1. Use houdini to create and publish all families

### CODE REVIEW

**Comment by MustafaJafar** [comment]
Will this PR be still useful if https://github.com/ynput/OpenPype/pull/5490 is merged ?

**Review by MustafaJafar — CHANGES_REQUESTED** [review]
Tested as a User these product types work fine
- camer
- imsage sequence
- ABC point cache
- BGEO point cache
- VDB cache
- Review

image

These families (that submits to farm) error 
- Mantra
- Karma

image

After applying this fix, it worked
https://github.com/ynput/OpenPype/blob/23e9f5e504c482bc13268d16190378142b5ca936/openpype/modules/deadline/plugins/publish/submit_houdini_render_deadline.py#L125
```python
        driver = instance.data["transientData"]["instance_node"]
```
image

**Comment by MustafaJafar** [comment]
I was Misled and thought that `instance.data["instance_node"]` is no longer exists after merging [Use host's node uniqueness for instance id in new publisher #5490](https://github.com/ynput/OpenPype/pull/5490), but I found out that it exists because of the following line: 
https://github.com/ynput/OpenPype/blob/2c7fc2160197e496a12bd23b41a0bebe7094688d/openpype/hosts/houdini/plugins/publish/collect_instances.py#L83

---

So, This PR 
1. Save `instnace_node` in `instance.data["transientData"]["instance_node"]` instead of `instance.data["instance_node"]`
2. removes `collect_instances.py` 

where `transientData` are saved ? and how `instance_node` value is maintained/updated ?

**Comment by MustafaJafar** [comment]
Let me rephrase my questions: 

- what are `transient data` and how are they stored ? 
- what will be `instance_node` value when the rop is duplicated ? in the light of the recently merged PR #5490

**Comment by BigRoy** [comment]
### Transient Data

Transient Data was once added because all instance data generated by the Creator had to be json serializable - the `hou.Node` is not and _transient data_ refers to the fact that it's runtime data that is really only valid for the current runtime session. You can't use that object instance after say, closing down houdini and opening it again.

It got added with this PR: https://github.com/ynput/OpenPype/pull/3897 where at first it was referred to as **lifetime data** but later got renamed to **transient data**.

It seems that _Transient Data_ unfortunately was never added to the documentation: https://openpype.io/docs/dev_publishing - probably best to create an issue about that.

---

> what will be instance.transient_data["instance_node"] value when the rop is duplicated ? in the light of the recently merged PR https://github.com/ynput/OpenPype/pull/5490

It will still refer to the ROP node, but by `hou.Node` instead of the node's string path. So it refers to the python `hou.Node` instance as opposed to a string path value.

It's likely however that some conflicts need to be resolved since that PR - but before anyone would do so I'd rather first know whether this PR is even remotely worth the effort and whether it's a good design choice.

As a reference, Fusion also uses the `transientData` and that's what it was initially added for so you knew exactly _what_ object you were referring to since just a node name/path there wasn't sufficient since Fusion allows to have multiple workfiles open at the same time in a single session. 🤯

**Comment by MustafaJafar** [comment]
> It seems that Transient Data unfortunately was never added to the documentation: https://openpype.io/docs/dev_publishing - probably best to create an issue about that.

yeah, I will add that to the community guide [AYON / Openpype Publish process - Development guide](https://community.ynput.io/t/ayon-openpype-publish-process-development-guide/734)

When I started learning about Openpype 4 months ago, I didn't expect all these details! 
and I love that!

**Comment by MustafaJafar** [comment]
That's pretty interesting! 
transient data: These data are not stored to scene and will be lost on object deletion.

I can recognize these three types of data, they are used in the Houdini creator. 
image


---
So, running this in Houdini's python source editor, can possibly tell us something. 
I don't know how it works but it's working, I tried opening old files and run the following script and it seems to catch old instances too. 

```python
from openpype.pipeline.create import CreateContext
from openpype.pipeline import registered_host

host = registered_host()
context = CreateContext(host, reset=True)

for instance in context.instances:
    print(instance.subset_name, instance.transient_data)
```


how Openpype is able to retrieve these transient data ?? 
and is it relevant to this node which I can't find ? 

image

**Comment by BigRoy** [comment]
The Creator collects the `transientData` - exactly as shown in this PR. Based on searching for the nodes relevant to the creator identifier the Creator assumes this is an instance node belonging to this creator - and then stores the `hou.Node()` in transient data, exactly like how previously the node's path was stored in the instance data.

Personally I see it like this: The _only_ reason for me that `hou.Node()` is in transient data is because it can't be converted to JSON data and thus by design it has to be put in `transientData`.

**_But_** transient data tells a bit more as well - it says 'do not store these on the node' - which happens to be the case for non-JSON data (because we can't serialize it well) but also for this instance node. We don't need to store it on the node, because it _is_ the node.

Not sure how to better explain it.

I know @iLLiCiTiT has a view of his own on it and likely can explain it better in a different way.

> I don't know how it works but it's working, I tried opening old files and run the following script and it seems to catch old instances too.

Exactly - the Creator already knows how to find the nodes, that logic never relied on the node's path being stored on the node, etc. All the Creator does was "find all nodes with attribute `creator_identifier` - if identifier matches this Creator class then it's an instance of this Creator. That's all how they are found. That hasn't changed.

The fact that `instance_node` was stored as a parm on the node was - in my view - a design flaw / bug from the beginning so it's good that since `instance_node` would be the actual node we're querying the parm from haha, so we don't need to store that data.

Similarly we can have `frameStart` and `frameEnd` instance data collected by the Creator but when _updating the instance_ **not** store the data in the JSON data parms but directly into the ROP nodes time parameters so that we frame start and end can both be set from the publisher AND on the Houdini ROP node. (This is sligthly more complex due to the fact that in Houdini the values can also be expression so the UI would have to likely show it as a `TextDef`, etc.) Anyway, I deviate too much. I hope this helps.

**Comment by MustafaJafar** [comment]
> The Creator collects the transientData - exactly as shown in this PR. Based on searching for the nodes relevant to the creator identifier the Creator assumes this is an instance node belonging to this creator - and then stores the hou.Node() in transient data, exactly like how previously the node's path was stored in the instance data.

I think I realized how it works, Thanks! 

So, there are two ways to create instances in Houdini 
- HoudiniCreator(...).create(...) # which creates nodes and instances from user input 
- HoudiniCreator(...).collect_instances()  # which create instances from existing nodes 

and when file opens, some plugin call  HoudiniCreator(...).collect_instances() which retrieves instances and their transient data

**Comment by MustafaJafar** [comment]
I like the idea introduced by this PR, I'm amazed that same approach is followed in Nuke and Fusion.

**Comment by BigRoy** [comment]
Would you just need me to resolve the conflicts, or are there other things to implement?

**Comment by MustafaJafar** [comment]
Nah, It's a casual PR revisiting during PRs cleaning up. 
what is the status of this PR :D ? and what's missing ?

**Comment by BigRoy** [comment]
Is it worth our while to refactor like this. Since it doesn't change anything end-user wise it doesn't bring any new features. So, biggest question is - do we **want** this? 

If so, I can resolve conflicts and prepare it for testing.

**Comment by MustafaJafar** [comment]
> So, biggest question is - do we **want** this?

I'm afraid I can't answer this question, I'll leave it to @moonyuet or @antirotor

File: openpype/hosts/houdini/plugins/publish/validate_no_errors.py
Comment (iLLiCiTiT) [reply openpype/hosts/houdini/plugins/publish/validate_no_errors.py:41]: I think this should be:
```python
instance_node = instance.data["transientData"].get("instance_node")
if instance_node is not None:
    validate_nodes.append(instance_node)
```

File: openpype/hosts/houdini/plugins/publish/collect_usd_layers.py
Comment (iLLiCiTiT) [reply openpype/hosts/houdini/plugins/publish/collect_usd_layers.py:59]:
```python
instance_node = instance.data["transientData"]["instance_node"]
for layer in usdlib.get_configured_save_layers(rop_node):
    ...
    layer_inst.data["instance_node"] = instance_node
    ...
```

File: openpype/hosts/houdini/plugins/publish/validate_vdb_output_node.py
Comment (iLLiCiTiT) [reply openpype/hosts/houdini/plugins/publish/validate_vdb_output_node.py:101]:
```
                instance_node = instance.data["transientData"]["instance_node"]
                formatting_data={
                    "message": message,
                    "rop_path": instance_node.path(),
                    "sop_path": output_path
                }
```

**Comment by iLLiCiTiT** [comment]
Is this PR still relevant? I think most of the changes are, but not sure if are up to date.

**Comment by mkolar** [comment]
Not sure how to deal with this one right now. Lot's of conflicts, but probably a reasonable change. @antirotor I might need your guidance here. 

Seems that it should be OP and AYON compatible out of the box though

**Comment by mkolar** [comment]
Because we're splitting OpenPype into ayon-core and individual host addons, this PR would have to be re-created to target one of those.

We're closing it down, but we'll he happy for a new PR to [ynput/ayon-core](https://github.com/ynput/ayon-core) or the host addon repository once it's up.

**Pull request closed without merging**
